File: 0752-open-the-lock/0752-open-the-lock.cpp
class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        unordered_set<string> visited;
        for(const auto &deadend : deadends){
            visited.insert(deadend);
        }

        string start = "0000";

        if(visited.find(start) != visited.end()) return -1;
        
        int moves = 0;
        queue<string> q;
        q.push(start);

        while(!q.empty()){
            int n = q.size();

            while(n--){
                string current = q.front(); q.pop();
                if(current == target) return moves;

                for(int i = 0; i < 4; i++){
                    char ch = current[i];

                    char increment = ch == '9' ? '0' : ch + 1;
                    char decrement = ch == '0' ? '9' : ch - 1;

                    current[i] = increment;
                    if(visited.find(current) == visited.end()){
                        visited.insert(current);
                        q.push(current);
                    }

                    current[i] = decrement;
                    if(visited.find(current) == visited.end()){
                        visited.insert(current);
                        q.push(current);
                    }

                    current[i] = ch;
                } 
            }
            moves++;
        }

        return -1;
    }
};

File: 0752-open-the-lock/0752-open-the-lock.java
class Solution {
    public int openLock(String[] deadends, String target) {
        Set<String> deadendSet = new HashSet<>(List.of(deadends));

        String start = "0000";

        if(deadendSet.contains(start)) return -1;
        if(target.equals(start)) return 0;

        Set<String> visited = new HashSet<>();

        Deque<String> q = new ArrayDeque<>();
        q.offer(start);
        visited.add(start);

        int moves = 0;

        while(!q.isEmpty()){
            int size = q.size();

            while(size-- > 0){
                String current = q.pop();

                if(target.equals(current)) return moves;

                if(deadendSet.contains(current)) continue;

                char[] currentArray = current.toCharArray();
                for(int i = 0; i < currentArray.length; i++){
                    char currentChar = currentArray[i];

                    //inrement with wrap-around
                    currentArray[i] = (char)(((((int)(currentChar - '0') + 1 + 10 ) % 10)) + '0');
                    String incrementString = new String(currentArray);
                
                    if(!visited.contains(incrementString)){
                        q.offer(incrementString);
                        visited.add(incrementString);
                    }
            
                    //decrement with wrap-around
                    currentArray[i] = (char)(((((int)(currentChar - '0') - 1 + 10) % 10)) + '0');

                    String decrementString = new String(currentArray);
                
                    if(!visited.contains(decrementString)){
                        q.offer(decrementString);
                        visited.add(decrementString);
                    }

                    currentArray[i] = currentChar;
                }
            }
            moves++;
        }

        return -1;
    }
}